url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/LaneAgent.java

```diff
@@ -0,0 +1,320 @@
+	public int extraRequestCount = 0;
+
+	public LaneStatus state;
+
+	String name;
+
+	public Semaphore animation = new Semaphore(0, true);
+
+	public class MyPart {
+		public Part part;
+		PartStatus status;
+
+		public MyPart(Part p) {
+			part = p;
+			status = PartStatus.BEGINNING_LANE;
+		}
+	}
+
+	public enum PartStatus {
+		BEGINNING_LANE, IN_LANE, END_LANE, TOLD_NEST, NEED_TO_DELIVER, DELIVERED
+	};
+
+	public enum LaneStatus {
+		FILLING, PURGING, WAITING, BROKEN, BROKEN_WHILE_PURGING;
+	};
+
+	FeederAgent feeder;
+	NestAgent nest;
+	LaneGraphics laneGUI;
+
+	public LaneAgent(String name) {
+		super();
+
+		this.name = name;
+		state = LaneStatus.FILLING;
+	}
+
+	@Override
+	public void msgINeedPart(PartType type) {
+		print("Received msgINeedPart");
+		requestList.add(type);
+		currentType = type;
+		stateChanged();
+	}
+
+	@Override
+	public void msgPurgeParts() {
+		print("Received msgPurgeParts");
+		if(state == LaneStatus.BROKEN_WHILE_PURGING || state == LaneStatus.BROKEN) {
+			state = LaneStatus.BROKEN_WHILE_PURGING;
+		} else {
+			state = LaneStatus.PURGING;
+		}
+		stateChanged();
+	}
+
+	@Override
+	public void msgHereIsPart(Part p) {
+		print("Received msgHereIsPart of type " + p.type.getName());
+		currentParts.add(new MyPart(p));
+		if (laneGUI != null) {
+			laneGUI.receivePart(p.partGraphics);
+		}
+
+		stateChanged();
+	}
+
+	@Override
+	public void msgGiveMePart() {
+		print("Received message give me part from nest");
+		synchronized (currentParts) {
+			for (MyPart p : currentParts) {
+				if (p.status == PartStatus.TOLD_NEST) {
+					p.status = PartStatus.NEED_TO_DELIVER;
+					break;
+				}
+			}
+		}
+		stateChanged();
+
+	}
+
+	@Override
+	public void msgReceivePartDone(PartGraphics part) {
+		print("Received msgReceivePartDone from graphics");
+		synchronized (currentParts) {
+			for (MyPart p : currentParts) {
+				if (p.status == PartStatus.BEGINNING_LANE) {
+					p.status = PartStatus.END_LANE;
+					break;
+				}
+			}
+		}
+		// animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgPurgeDone() {
+		print("Received msgPurgeDone");
+		animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgGivePartToNestDone(PartGraphics part) {
+		print("Received msgGivePartToNestDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+	
+	@Override
+	public void msgChangeAmplitude() {
+		print("Received msgChangeAmplitude");
+		laneGUI.unjam();
+	}
+	
+	@Override
+	public void msgFixYourself() {
+		if(state == LaneStatus.BROKEN_WHILE_PURGING) {
+			state = LaneStatus.PURGING;
+		}
+		else {
+			state = LaneStatus.FILLING;
+		}
+		stateChanged();
+	}
+
+	@Override
+	public void msgBreakThis() {
+		print("Received message msgBreakThis");
+		if(state == LaneStatus.PURGING) {
+			state = LaneStatus.BROKEN_WHILE_PURGING;
+		}
+		else { 
+			state = LaneStatus.BROKEN;
+		}
+		stateChanged();
+	}
+	
+	@Override
+	public boolean pickAndExecuteAnAction() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 181-227 into a separate method.